숫자게임 [STEP 2] seohyeon2, ilim

seohyeon2 opened the pull request:
# 숫자게임 [STEP 2] seohyeon2, ilim

안녕하세요 도니! @westeastyear
seohyeon2, ilim 입니다.
숫자 게임 STEP2 PR 보냅니다.
조언받고 싶은 부분과 순서도를 올리니 확인 부탁드리며, 
이번 리뷰도 잘 부탁드립니다🙇🏻‍♀️

## 고민했던 점, 조언 받고 싶은 부분
- 이번 과제에서, 정규식을 사용하여 사용자가 입력한 숫자의 유효성 검증을 하는 걸 지양하시는지 궁금합니다. 숫자 입력값의 조건이 명확하다고 생각하여 정규식을 사용했습니다. 숫자 3개를 공백으로 구분한 입력값 외에 유효성 검증을 통과하는 입력값은 없다고 판단했기 때문입니다.
- 클로저의 축약 형식도 가독성이 좋다고 생각하시는지 궁금합니다. compactMap 사용 시, 클로저를 축약한 형태로 코드를 작성했습니다(main.swift 70번째). 인라인 클로저와 축약 인자 사용도 축약하지 않은 상태만큼 가독성이 괜찮다고 느끼시는지 궁금합니다.

## 순서도
![step1,2 flowchart](https://user-images.githubusercontent.com/50102522/163167155-6d6d4166-f302-46d8-a7ac-419b6d152e0e.png)

seohyeon2 committed:
feat: 사용자 메뉴를 출력하고 입력받는 함수 구현

seohyeon2 committed:
feat: 에러메세지 출력 함수 추가

Commit:
feat: 숫자 3개 입력받는 함수 작성

seohyeon2 committed:
feat: 유효성 검사하는 함수 구현

Commit:
feat: verifyInput 함수 기능 수정

seohyeon2 committed:
feat: 정규식을 이용해 사용자 숫자를 검증하는 함수 생성

Commit:
refactor: 사용자 입력 기능 수정
1. 게임을 지속적으로 실행하는 반복문 위치 변경
2. 게임 시행 횟수 변수 위치 변경
3. 게임 시행 횟수 차감 코드 위치 변경

seohyeon2 committed:
docs: STEP2를 하며 느꼈던 점들을README에 추가함

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,67 @@ func printResult(strikeCount: Int, ballCount: Int) {
+    while totalTrialNumber > 0 {
+        print("숫자 3개를 띄어쓰기로 구분하여 입력해주세요.")
+        print("중복 숫자는 허용하지 않습니다.")
+        print("입력 : ", terminator: "")
```
`Swift`에서는 `MultiLine String`을 지원한답니다. 한번 적용해 볼까요?

seohyeon2 replied:
적용하여 수정했습니다!

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,67 @@ func printResult(strikeCount: Int, ballCount: Int) {
+    while isStart {
+        print("1. 게임시작")
+        print("2. 게임종료")
+        print("원하는 기능을 선택해주세요 : " , terminator: "")
```
이 부분도 `MultiLine String`을 적용해 보세요!

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,67 @@ func printResult(strikeCount: Int, ballCount: Int) {
+    }
+}
+
+func printErrorMessage() {
+    print("입력이 잘못되었습니다")
+}
```
함수로 빼신점 아주 칭찬합니다👍🏻👍🏻

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,67 @@ func printResult(strikeCount: Int, ballCount: Int) {
 }
 
-startGame()
+func inputNumbers() {
```
주체가 누군지 알려주는 뜻이 들어갔으면 좋겠어요!

iilim replied:
해당 게임 프로그램에서 숫자를 입력한다 라는 경우의 수는 플레이어밖에 없습니다.
왜냐하면 컴퓨터의 난수는 자동으로 생성되기 때문입니다
그래서 숫자를 입력하는 경우의 수는 자연스럽게 플레이어밖에 없기 때문에
주체를 명시하지 않아도 그 뜻이 명확하게 전달 된다고 생각하는데, 그래도 굳이 바꿔야 할까요?

westeastyear replied:
이렇게 작성한 의도를 따로 설명을 하지 않아도, 이 코드를 처음 읽는 사람이 알수 있게 하는 것이 좋은 네이밍이라고 생각이 듭니다.
여러분의 생각은 어떠신가요? 주체에 관한 이야기는 저의 개인적인 의견이였을 뿐입니다ㅎㅎ

iilim replied:
처음에는 굳이 바꿀 필요가 있나 라고 생각했는데 피드백 보니 그말도 맞는 것 같습니다.... 다음부터는 핑계 대지 않고 그냥 바로 바로 바꾸겠습니다!

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,67 @@ func printResult(strikeCount: Int, ballCount: Int) {
+    printResult(strikeCount: strikeCount, ballCount: ballCount)
+}
+
+func verifyWithRegularExpression(playerInput: String) -> Bool {
+    if ((playerInput.range(of: #"^[1-9]\s[1-9]\s[1-9]$"# ,options: .regularExpression)) != nil) {
+        return true
     }
+    return false
```
정규표현식을 사용하셨네요🤭🤭
정규표현식을 사용한 55번 라인을 글로 설명해볼 수 있을까요? 
그리고`!= nil`과 같은 기능을 할 수 있는 코드는 무엇일까요? 부정의 부정은 코드를 읽기 어렵게 만듭답니다.😅

seohyeon2 replied:
앗.. 옵셔널 바인딩..으로 처리를 안했네요😅

seohyeon2 replied:
55번째라인은 of 라인안에 들어가 있는 패턴이 playerInput으로 들어온 값이랑 일치하는지 검증하는 코드입니다.
검증을 시작(^)하고 1~9 사이에 숫자가 1개 들어가는 가 ([1-9]),바로 뒤에 공백이 오는가(\s)를 물어보고 검증을 종료($)하는 코드입니다.😊

이 상태에서 바로 옵셔널 바인딩으로 처리를 하려고 보니, Xcode에서 55번째와 같은 형태로 fix하라는 경고문이 떴습니다.🥲
그래서 이 코드를 따로 상수에 넣어주고 nill 값과 비교하여 부정의 부정 코드를 수정했습니다.

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,67 @@ func printResult(strikeCount: Int, ballCount: Int) {
-    print("임의의 수 : ", terminator: "")
-    for randomNumber in playerRandomNumbers {
-        print(randomNumber, terminator: " ")
+func startGame(playerNumbers: Array<Int>) {
+    computerRandomNumbers = makeRandomNumbers()
+    let (strikeCount, ballCount) = checkStrikeBall(playerRandomNumbers: playerNumbers)
+    printResult(strikeCount: strikeCount, ballCount: ballCount)
```
`startGame`이라는 이름과 함수의 내용이 어울리지 않는다고 생각됩니다. 이 메서드는 어떤 기능을 하는 함수인가요?

seohyeon2 replied:
게임을 시작하고 전반적인 과정을 처리하는 함수로 작성했습니다.
도니의 말처럼 어울리지 않는 듯 하여 '과정을 처리한다'는 역할에 맞추어  `processGame`이라 변경했습니다!

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,67 @@ func printResult(strikeCount: Int, ballCount: Int) {
         totalTrialNumber -= 1
-        let (strikeCount, ballCount) = checkStrikeBall(playerRandomNumbers: playerRandomNumbers)
-        printResult(strikeCount: strikeCount, ballCount: ballCount)
+        startGame(playerNumbers: playerNumbers.compactMap { Int($0) })
```
고차함수 `compactMap`을 사용하셨네요👍🏻👍🏻 
이 친구는 어떤 기능을 가지고 있나요?
축약을 하지 않고 작성한다면 어떤 모습일까요?

seohyeon2 replied:
축약을 하지 않고 바로 넣으려고 보니 에러가 많이 뜨더라구요..
아마 축약하지 않은 형태를 잘 못 구현한 거 같아요😂
제대로 알지 못하는 개념을 코드로 구현하는 건 아닌 듯 싶어,
compactMap과 비슷한 기능을 하는 함수(convertStringArrayToIntArra)를 만들어 처리하였습니다!

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,67 @@ func printResult(strikeCount: Int, ballCount: Int) {
+    }
+    if isCorrect {
+        let playerNumbers = playerInput.components(separatedBy: " ")
         totalTrialNumber -= 1
```
팀원과 README에 작성한 것처럼, 함수가 단일 기능을 가지도록 노력해 봅시다! 한번 홀쭉하게 만들어 보아요😆

iilim replied:
함수의 단일기능이라는 개념이 살짝 혼란인 와서 질문드립니다
만약 함수안에서 다른 2개의 함수를 호출하는 경우 이런건 단일기능이 아니라 2개의 기능을 하는 함수인가요?
또한 2개의 병렬관계를 이루는 조건문이 존재하며 각각의 조건문의 조건들이 서로 전혀 상관없는 독립적이 관계를 이루는 경우,
이 또한 단일기능이 아니라 2개의 기능을 담당하는 함수로 보시는건지가 궁금합니다!
하나의 함수는 하나의 일만 해야한다라는 개념이 처음엔 들었을때 직관적으로 이해되는 것처럼 느껴지는데
막상 코딩을 하고나서 코드를 가만히 살펴보면 이건 단일기능을 하는 함수가 맞나 싶은 생각을 꽤 많이 하게 됩니다

westeastyear replied:
게임 첫 시작의 경우, 분기를 나눠야 한다면 그 곳에서는 2-3개의 함수를 호출할 수 있을 것 같아요.
그 안에서 호출되는 함수는 단일 기능을 하는 게 좋겠죠?!
무조건 하나의 함수에는 하나의 기능이라기보다는, 이렇게 했을 때 코드가 더 깔끔해지고, 충분한 이유가 있다면 2개 호출도 괜찮다고 생각합니다 🙂
가독성 있는 코드를 짜는 게 목적이 되면 좋을 것 같아요!

westeastyear replied:
야곰의 코멘트도 추가합니다!!

한가지 일 이라는 단위가 중요할 것 같은데요, 
우리가 한가지 일을 어떻게 단위를 나누느냐가 핵심인 것 같아요.
만약에 저녁상을 차린다를 하나의 일이라고 하면
하위 일로
재료 준비하기 / 요리하기 / 수저 놓기 / 가족 부르기 
등으로 하위 일이 생길거고
재료 준비하기 라는 일은
마트 가기 / 재료 고르기 / 계산하기 / 집으로 운반하기
등으로 나눌 수 있겠죠
이렇게 치면 한가지 일이라는 것은 다양한 관점으로 볼 수 있겠죠?
한가지 일을 하위의 여러가지 일로 나눌 수 있고 그것을 각각의 함수로 나눌 수 있고요,
가장 큰 상위의 일은 더 세부적인 함수 여러개의 호출로 볼 수도 있는거죠.

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,67 @@ func printResult(strikeCount: Int, ballCount: Int) {
+    }
+}
+
+func showMenu() {
```
어떤 메뉴를 보여주는 걸까요? 🤔🤔

seohyeon2 replied:
게임의 시작과 종료의 메뉴를 보여주는 함수여서 showGameMenu()로 수정했습니다..!
혹시 더 좋은 네이밍이 있다면 추천해주세요😌

westeastyear replied:
충분합니다👍🏻

westeastyear reviewed (commented):
안녕하세요! @seohyeon2 @iilim 👋🏻
빠르게 [STEP 2]를 수행하셔서 놀랬답니다ㅎㅎ
다들 고생하셨습니다.

`이번 과제에서, 정규식을 사용하여 사용자가 입력한 숫자의 유효성 검증을 하는 걸 지양하시는지 궁금합니다. 숫자 입력값의 조건이 명확하다고 생각하여 정규식을 사용했습니다. 숫자 3개를 공백으로 구분한 입력값 외에 유효성 검증을 통과하는 입력값은 없다고 판단했기 때문입니다.`
=> 초반부터 정규식이 나올줄은 꿈에도 몰랐습니다ㅎㅎ 두분이 충분히 상의를 하고 두분다 코드를 이해하여 작성하였다면, 아주 좋은 방법이고 좋은 자세라고 생각됩니다. 코드엔 정답이 없으니까요!!

`클로저의 축약 형식도 가독성이 좋다고 생각하시는지 궁금합니다. compactMap 사용 시, 클로저를 축약한 형태로 코드를 작성했습니다(main.swift 70번째). 인라인 클로저와 축약 인자 사용도 축약하지 않은 상태만큼 가독성이 괜찮다고 느끼시는지 궁금합니다.`
=> 두분은 축약하지 않은 상태로도 작성이 가능하실까요? 가독성도 물론 중요하지만 캠프에서는 코드를 좀더 파헤쳐볼수는 시간이라고 생각됩니다. 나중에 현업에서는 동료와 충분한 이야기를 나눈후 축약규칙을 정한다고 알고 있습니다! 둘다 할 줄 알면 당연히 좋겠죠?ㅎㅎ

## 😆 네이밍
함수, 변수에 역시 신경을 많이 써주셨네요.
특히 `Bool` 변수 네이밍에 눈길이 많이 갔습니다. 👏👏

## 🤔 순서도에 대한 리뷰
피드백대로 `README`와 `PR`에 적용하신점 칭찬합니다👍🏻👍🏻
하지만, 코드와 순서도를 같이 보면서 진행하니 아쉬운 부분들이 보였습니다.
코드 수정후 순서도도 같이 수정해 볼까요? [STEP 2]를 기반으로 좀더 디테일하게 하면 좋을거 같습니다.😊

## 😭 사용자 승리...?😫
코드를 다 작성하시고 `사용자 승리`를 확인하셨을까요?? 
제가 확인한 바로는 컴퓨터를 전혀 이길 수가 없었습니다...! 😭
왜 그런지 찾아보시고 코멘트와 함께 로직 수정 부탁드립니다.
<img src="https://user-images.githubusercontent.com/74251593/163459950-5184cc79-b68c-491a-b8e9-5435ba67a47a.png" width="60%">

## 🥳 README 🎉
피드백대로 `README`를 업데이트 하고 계시네요👍🏻👍🏻
나중에 시간이 지나면 내가 뭘했는지 기억이 나지 않는답니다. 그래서 제가 눈물을 흘리고 있죠...🥲
여러분이 작성하고 기록하는 모든것들이 다 포트폴리오가 될꺼에요!
귀찮다고 미루면 미래의 내가 힘들어져요~🤣

Commit:
feat: verifyInput 함수 기능 수정
verifyInput 함수에 있던 조건문 2개를 함수화 시켜서 새로운 함수 2개 생성

Commit:
refactor: 함수 이름 변경(playGame -> processGame)

seohyeon2 committed:
docs: 순서도 사진 변경 및 해결 부분 추가

seohyeon2 committed:
fix: totalTrialNumber 값 초기화 안되는 이슈 수정

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,91 @@ func printResult(strikeCount: Int, ballCount: Int) {
-        totalTrialNumber -= 1
-        let (strikeCount, ballCount) = checkStrikeBall(playerRandomNumbers: playerRandomNumbers)
-        printResult(strikeCount: strikeCount, ballCount: ballCount)
+        print("""
+        숫자 3개를 띄어쓰기로 구분하여 입력해주세요.
+        중복 숫자는 허용하지 않습니다.
+        입력 : \
+        
+        """, terminator: "")
```
`MultiLine`으로 바꾸셨군요👍🏻 👍🏻  앞으로도 잘 활용해 볼 수 있겠죠?😊
다만 99번, 100번 라인에 불필요한게 있네요! 
불필요한 걸 삭제후, `terminator`에 무언가를 추가하면 해결할 수 있답니다! 무엇이 필요할지 고민해 볼까요?

iilim replied:
아 ㅋㅋㅋ..... 피드백 보자마자 생각났습니다
ㅋㅋㅋㅋ 어이가 없네요

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,91 @@ func printResult(strikeCount: Int, ballCount: Int) {
+        2. 게임종료
+        원하는 기능을 선택해주세요 : \
+        
+        """, terminator: "")
```
이 부분도 마찬가지이네요? 한번 리펙토링 해봅시다~ 😉

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,91 @@ func printResult(strikeCount: Int, ballCount: Int) {
+        
+        """, terminator: "")
+        let inputMenuNumber = readLine() ?? ""
+        switch inputMenuNumber {
+        case "1" :
+            computerRandomNumbers = makeRandomNumbers()
+            inputNumbers()
+        case "2" :
+            isStart = false
+            break
+        default:
+            printErrorMessage()
+        }
```
너무 자연스러워서 이제야 발견했네요 🙇🏻‍♂️ 🙇🏻‍♂️
`switch`구문을 사용하신점에 칭찬 한스푼 첨가합니다! 🥄

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,91 @@ func printResult(strikeCount: Int, ballCount: Int) {
+
+func verifyWithRegularExpression(playerInput: String) -> Bool {
+    let playerInputAfterVerification = playerInput.range(of: #"^[1-9]\s[1-9]\s[1-9]$"# ,options: .regularExpression)
+    if playerInputAfterVerification == nil {
+        return false
+    }
```
전보다 훨신 읽기 편하네요 👍🏻👍🏻 
아주 좋아요 👏👏

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,91 @@ func printResult(strikeCount: Int, ballCount: Int) {
+func checkDuplicateNumber(playerInput: String) -> Bool {
+    var isCorrect = false
+    let playerInputWithoutWhitespace = playerInput.replacingOccurrences(of: " ", with: "")
+    if Set(playerInputWithoutWhitespace).count == playerInputWithoutWhitespace.count {
```
64번 라인의 코드를 자세히 설명 부탁드립니다!!

iilim replied:
playerInput은 "정수 정수 정수" 형식의 문자열입니다. 
따라서 63번 라인을 실행하면 공백이 제거된 문자열이 상수에 저장되고
그 상수를 set에 넣고나서 count와 넣지 안은 상수의 길이를 비교하는 것입니다
왜냐하면 중복된 원소를 체크하는 함수이기 때문에
만약 중복되는 원소가 없다면 set에 넣었어도 길이의 변화는 없기 때문입니다
만약 중복되는 원소가 있다면 기존 문자열의 길이와 set에 넣은 후의 길이가 달라지게 됩니다!
그런 의미로 작성한 코드입니다 이해가 되셨을까요? 설명이 너무 장황한것 같아서....

iilim replied:
해당 방법이 애플이 공식적으로 권장하는 중복을 체크하는 방법이라는 글을 보고 적용해봤습니다

westeastyear replied:
저도 한수 배워갑니다 🙇🏻‍♂️🙇🏻‍♂️

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,91 @@ func printResult(strikeCount: Int, ballCount: Int) {
-    print("임의의 수 : ", terminator: "")
-    for randomNumber in playerRandomNumbers {
-        print(randomNumber, terminator: " ")
+func processGame(playerNumbers: Array<Int>) {
+    let (strikeCount, ballCount) = checkStrikeBall(playerRandomNumbers: playerNumbers)
+    printResult(strikeCount: strikeCount, ballCount: ballCount)
```
`processGame` 보다는 `strike`와 `ball`에 관련된 이름은 어떨까요? 함수명이 내용과 상이하다고 느껴지네요🤔
조금더 고민해볼까요?

iilim replied:
이거 한참 고민하다가 process로 정한거였는데.... 다시 고민하겠습니다

westeastyear commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -44,24 +45,91 @@ func printResult(strikeCount: Int, ballCount: Int) {
+    print("입력이 잘못되었습니다")
+}
 
+showGameMenu()
```
🥳🥳🥳

westeastyear reviewed (commented):
함수들이 많이 홀쭉해졌네요! 고생하셨습니다! 👏👏

언제나 변수, 함수명은 개발자의 숙명과도 같습니다.
변수, 함수명에는 타입(ex: Int, Array)의 이름 사용을 지양한다는걸 알아두시면 좋겠습니다.(수정하지 않으셔도 됩니다ㅎㅎ)
하지만, 이름에 타입을 작성해주는 경우도 있답니다! 찾아보시면 공부가 되겠죠? 

순서도도 코드와 똑같이 흘러가게 잘 그려주셔서 확인하기 편하였고, 피드백을 잘 수용하고 있다는걸 느낄 수 있었답니다! 👍🏻👍🏻 

나머지는 코멘트를 확인해주세요~🧑🏻‍💻 👩🏻‍💻

Commit:
refactor: multiline string terminator, 함수이름 수정(processGame -> checkAndPrintStrikeBall)
1. 기존의 multiline string의 불필요한 코드 수정, terminator로 공백 처리
2. 두가지 함수의 기능을 모두 담은 함수 이름으로 수정
